Initial modification for clamav server to handle REST request

rainqin opened the pull request:
Modified the clamav server.
- It doesn't need to config the bucket before scanning.
- It scans a single file with format:
'{"kind": "storage#object","name":"<FIile name>", "bucket": "<Bucket name>" }'
- It doesn't move the file after scanning.
- The response is in Json with message "clean" "infected" or "not exists"

rainqin committed:
Modified clamav server for REST api access.

Modified the clamav server.
- It doesn't need to config the bucket before scanning.
- It scans a single file with format:
'{"kind": "storage#object","name":"<FIile name>", "bucket": "<Bucket name>" }'
- It doesn't move the file after scanning.
- The response is in Json with message "clean" "infected" or "not exists"

rainqin committed:
Add config.json to the respository as well

Add config.json to the respository as well

The origin code base ignores the config.json, which is a must for server starting.

ricebin commented on cloudrun-malware-scanner/config.json:
```diff
@@ -9,12 +9,5 @@
-    }
-  ],
-  "ClamCvdMirrorBucket": "cvd-mirror-bucket-name"
+  "ClamCvdMirrorBucket": "cvd-mirror-development-344423"
```
why is this set to cvd-mirror-development-344423?

rainqin replied:
The name can be anything. By default, it's cvd-mirror-$PROJECT_ID
This bucket is used to maintain a local mirror of the malware definitions database, which prevents rate limiting from being triggered by the ClamAV CDN.
I haven't setup this pipeline yet.

ricebin replied:
i am saying we cant have a default here
it can just take the bucket from the input request
or on server boot up (maybe env var, or startup arg)

rainqin replied:
Could you tell more about why we can't have a default here?
This bucket is created by me. It's just some bucket to hold the mirror of the db.

rainqin replied:
Let's have a short sync tomorrow (Friday) to clarify this.

rainqin replied:
Done. Put the config into environment variables

ricebin replied:
i would still like to remove this default
so that if CVD_MIRROR_BUCKET is not set, this service does not boot

rainqin replied:
Yes, it is like this now.
if CVD_MIRROR_BUCKET is not set, this service does not boot

ricebin replied:
oh.. then i am confused what this is for then

rainqin replied:
Server won't read from this config file. I can remove this for readability, it doesn't affect functionality
When docker start up, it will read the environment variable and overwrite this json value.
Other clamav config files will generate base on this value.

ricebin replied:
oh.. then can we just remove this?

ricebin commented on cloudrun-malware-scanner/server.js:
```diff
@@ -131,21 +130,14 @@ async function handleGcsObject(req, res) {
       // Warn but return successful to client.
       logger.warn(`Ignoring no longer existing file: gs://${file.bucket}/${file.name}`);
-      res.json({status: 'deleted'});
+      res.json({status: 'not exists'});
```
can u return an actual error status?

rainqin replied:
404 sg to u?

ricebin replied:
404 is fine

rainqin replied:
Done.

ricebin commented on cloudrun-malware-scanner/server.js:
```diff
@@ -167,26 +159,18 @@ async function handleGcsObject(req, res) {
         file.name}: INFECTED ${result} (${
         fileSize} bytes in ${scanDuration} ms)`);
-      metrics.writeScanInfected(config.unscanned, config.quarantined, fileSize,
+      metrics.writeScanInfected("gs://${file.bucket}/${file.name}", fileSize,
```
u cant have file.name in the metrics 
the cardinality is too high

rainqin replied:
That's true. Do we want to separate the metric in bucket? or just put them together?

ricebin replied:
by bucket is fine for now

rainqin replied:
Done.

ricebin commented:
update the pr title to something more meaningful if possible

rainqin committed:
Resolve comments.

While file not found, return 404 instead of OK.
Metrics based on buckets only.

rainqin committed:
Set up clamav-scanner-mirror-update pipeline

Set up clamav-scanner-mirror-update pipeline.
The pipeline runs every 2 hours to update the latest clamav database to cvd-mirror-development-344423

rainqin committed:
Put the bucket config in the environment variables

The clamav depends on several config files, such as /etc/clamav/clamd.conf,
/etc/clamav/freshclam.conf

So I still keep the config.json file, but instead of reading from the config.json, I read from environment variables and write to the config.json for the mirror bucket. This is to avoid making big changes to the clamav underlying logic.

ricebin commented on cloudrun-malware-scanner/server.js:
```diff
@@ -43,29 +43,7 @@ const CLAMD_TIMEOUT = 600000;
 // Note scanning a 500MiB file can take 5 minutes, so ensure timeout is
 // large enough.
 const MAX_FILE_SIZE = 500000000; // 500MiB
```
we are going to need increase this .. maybe make it into env var as well
some of the games are on the order of x gigs

ricebin commented on cloudrun-malware-scanner/server.js:
```diff
@@ -131,21 +109,17 @@ async function handleGcsObject(req, res) {
       // Warn but return successful to client.
       logger.warn(`Ignoring no longer existing file: gs://${file.bucket}/${file.name}`);
-      res.json({status: 'deleted'});
+      handleErrorResponse(
```
can u fix line 96 and line 100 and line 105 to return actual errors when request is invalid?
something like 400 error code?
